versioning/BRM/extentmap.cpp

```diff
@@ -269,19 +270,306 @@ FreeListImpl::FreeListImpl(unsigned key, off_t size, bool readOnly) :
 {
 }
 
+/*static*/
+std::mutex ExtentMapIndexImpl::fInstanceMutex_;
+
+/*static*/
+ExtentMapIndexImpl* ExtentMapIndexImpl::fInstance_ = nullptr;
+
+/*static*/
+ExtentMapIndexImpl* ExtentMapIndexImpl::makeExtentMapIndexImpl(unsigned key, off_t size, bool readOnly)
+{
+    std::lock_guard<std::mutex> lock(fInstanceMutex_);
+
+    if (fInstance_)
+    {
+        if (size != fInstance_->getShmemSize())
+        {
+            fInstance_->fBRMManagedShmMemImpl_.remap();
+        }
+
+        return fInstance_;
+    }
+
+    fInstance_ = new ExtentMapIndexImpl(key, size, readOnly);
+    fInstance_->createExtentMapIndexIfNeeded();
+
+    return fInstance_;
+}
+
+ExtentMapIndexImpl::ExtentMapIndexImpl(unsigned key, off_t size, bool readOnly) :
+    fBRMManagedShmMemImpl_(key, size, readOnly)
+{
+}
+
+void ExtentMapIndexImpl::createExtentMapIndexIfNeeded()
+{
+    // pair<T*, size>
+    // Hide fShmSegment
+    auto managedShmemSearchPair = fBRMManagedShmMemImpl_.getManagedSegment()->find<ExtentMapIndex>(EmIndexObjectName);
+    if (!managedShmemSearchPair.first || managedShmemSearchPair.second == 0)
+    {
+        ShmVoidAllocator alloc(fBRMManagedShmMemImpl_.getManagedSegment()->get_segment_manager());
+        fBRMManagedShmMemImpl_.getManagedSegment()->construct<ExtentMapIndex>(EmIndexObjectName)(alloc);
+    }
+}
+
+ExtentMapIndex* ExtentMapIndexImpl::get()
+{
+    // pair<T*, size>
+    // Hide fShmSegment
+    auto managedShmemSearchPair = fBRMManagedShmMemImpl_.getManagedSegment()->find<ExtentMapIndex>(EmIndexObjectName);
+    assert(managedShmemSearchPair.first && managedShmemSearchPair.second > 0);
+    return managedShmemSearchPair.first;
+}
+
+bool ExtentMapIndexImpl::growIfNeeded(const size_t memoryNeeded)
+{
+    auto freeShmem = getShmemFree();
+    // Worst case managed segment can't get continues buffer with len = memoryNeeded
+    if (freeShmem < memoryNeeded)
+    {
+        const size_t currentShmemSize = getShmemSize();
+        constexpr static const size_t minAllowance = 16 * 1024 * 1024;
+        const size_t newShmemSize = std::max(minAllowance, memoryNeeded) + currentShmemSize;
+        grow(newShmemSize);
+        return true;
+    }
+    return false;
+}
+
+InsertUpdateShmemKeyPair ExtentMapIndexImpl::insert(const EMEntry& emEntry, const size_t emIdx)
+{
+    auto dbRoot = emEntry.dbRoot;
+    assert(dbRoot > 0 && dbRoot <= numeric_limits<uint64_t>::max());
+    auto* extentMapIndexPtr = get();
+    assert(extentMapIndexPtr);
```

Comment:
Nit, seems like we have an assert in `get` method.

Reply:
Good point.